### PR TITLE
Added assertions for runtime Builder, changed create_blocking_pool interface

### PR DESCRIPTION
**Pull request by olegnn**
## Motivation

If user writes code like 
```rust
let runtime = Builder::new().core_threads(4).build().unwrap();
```
, he expects that runtime will have 4 threads, however currently this argument will be silently unused.

Moreover, if runtime isn't configured with `threaded_scheduler`, we don't need to call `loom::sys::num_cpus` function at all, but now it will be called for every runtime anyway.

## Solution

This PR adds assertion for this case and calls `loom::sys::num_cpus` only in case if runtime is configured with `threaded_scheduler`.

### CODE REVIEW

**Comment by vorot93** [comment]
I would strongly consider removing `basic_scheduler` and `threaded_scheduler` knobs at all. Number of core threads should be the sole determinant as to which scheduler to choose, at least with the current dichotomy.

**Comment by sfackler** [comment]
@vorot93 Doesn't a basic scheduler behave differently than a threaded scheduler with 1 core thread? In particular, a threaded scheduler will continue to process spawned tasks even when the Runtime is not running a future via block_on.

**Comment by carllerche** [comment]
@sfackler I've been thinking about this.... I will try to write up thoughts in more detail sometime (before making a change).

tl;dr, i think we *shoudl* get rid of `basic_scheduler` and `threaded_scheduler`. When setting `core_threads(1)` it would spawn a new thread w/ the basic scheduler implementation. To get the *current* `basic_scheduler` behavior, you would specify `core_threads(0)`.

The downside is that `core_threads(0)` is a bit "hidden" but i _think_ that is OK as it is a more "advanced" use case. Anyway, I will need to write up these thoughts and open to discussion. It would be an 0.3 change.

**Comment by sfackler** [comment]
Ah sure, core_threads(0) seems more plausible than core_threads(1).

**Comment by carllerche** [comment]
Unfortunately, this change will result in potential panics in people's existing applications. The question is whether or not this is acceptable in a patch release.

Yes, if there is a panic, they are *probably* not getting what they are expecting, but it is unclear to me if this should be fixed in 0.2 as a patch or in 0.3 as an overhaul to the configuration system.

Thoughts @tokio-rs/maintainers ?

**Comment by dekellum** [comment]
> Unfortunately, this change will result in potential panics in people's existing applications. The question is whether or not this is acceptable in a patch release.

Sounds like no, by definition.

**Comment by olegnn** [comment]
@carllerche maybe use `debug_assert` instead? Then existing applications running in `release` mode won't panic, but this will alert developers that they're doing wrong things.

**Comment by LucioFranco** [comment]
I vote we do this in 0.3 with an overhaul :+1:

**Comment by Darksonn** [comment]
I will close this for now as it probably won't look exactly like this when we decide what to do for v0.3. Progress is tracked by #2698.
